### DISCUSSION TITLE
Fit one image into another

### DISCUSSION THREAD

**Discussion by Jay123210599** [ImageMagick/ImageMagick#7430] [Help]
### ImageMagick version

7.1.1-27

### Operating system, version and so on

Windows 11

### Description

I'm trying to fix this image into this template. How do I readjust the size so that it fits?
![time to fit](https://github.com/ImageMagick/ImageMagick/assets/141977822/dcbc4a1f-8523-4efc-bd33-0bb8f6669ffb)
![template](https://github.com/ImageMagick/ImageMagick/assets/141977822/c55e0be5-0d36-4b1d-9742-0907846695fa)

**Comment by fmw42** [comment]
Where do you want to insert it?  Do you want the image to fill your whole template or only within the white rectangle?

Either way, your image aspect is landscape and the template and the white square are square or portrait.  So if you force a fit of the full image, it will be distorted.  So you have to (center) crop your input to the same aspect as template or white square.  Is that OK?

**Reply by Jay123210599** [reply]
within the white rectangle, and yes, so long as I don't lose parts of my input for the output.

**Reply by fmw42** [reply]
As I said above the process will have to crop the center of your input to match the aspect of your white square in the template or distort the image to fit the square.  Which do you want?

**Reply by Jay123210599** [reply]
If cropping doesn't involve the whole image into the white rectangle, then distort it.

**Comment by fmw42** [comment]
Here are two solutions.  The first crops to the same aspect and resizes your input to fill the white square and the second just resizes and pads with white to fill the white square.

Unix syntax

```
# get dimensions of white box using connected components
cropbox=`im7 magick template.png \
-colorspace gray \
-auto-threshold otsu \
-type bilevel \
-define connected-components:verbose=true \
-define connected-components:mean-color=true \
-define connected-components:keep-top=2 \
-define connected-components:exclude-header=true \
-connected-components 4 null: | grep "gray(255)" | awk '{print $2}' | tr "x" "+"`

# get aspect from width and height and offsets
ww=`echo $cropbox | cut -d+ -f1`
hh=`echo $cropbox | cut -d+ -f2`
ox=`echo $cropbox | cut -d+ -f3`
oy=`echo $cropbox | cut -d+ -f4`
aspect="${ww}:${hh}"

# center crop the input and resize to dimensions of white box
im7 magick template.png +repage \
\( img.png +repage -crop $aspect +repage -resize ${ww}x${hh} \) \
-gravity northwest -geometry +${ox}+${oy} -compose over -composite \
result1.png

# alternate: resize the input and pad with white to fill to the dimensions of the white box
im7 magick template.png +repage \
\( img.png +repage -resize ${ww}x${hh} -background white -gravity center -extent ${ww}x${hh} +repage \) \
-gravity northwest -geometry +${ox}+${oy} -compose over -composite \
result2.png

# alternate2: resize the input distorting to fill to the dimensions of the white box
im7 magick template.png +repage \
\( img.png +repage -resize ${ww}x${hh}! \) \
-gravity northwest -geometry +${ox}+${oy} -compose over -composite \
result3.png

```

Result1:

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/84209860-c396-47b9-8fd0-e504b662444d)


Result2:

![result2](https://github.com/ImageMagick/ImageMagick/assets/13458218/23922f16-3729-4d9a-948b-b57400887737)


Result3:

![result3](https://github.com/ImageMagick/ImageMagick/assets/13458218/b2ef7e1e-a856-438b-8641-18f043294e9a)

**Reply by Jay123210599** [reply]
For the second solution, how do I resize the image to fill in the white space?

**Reply by fmw42** [reply]
See my answer. I added result3 to distort the input to fill the white box.

**Reply by fmw42** [reply]
Note: my solution is for Unix.  You can run it under Windows 10/11 unix or get Cygwin.  Otherwise, one of the Windows users will need to translate to DOS Windows bat file syntax.

**Reply by Jay123210599** [reply]
How do I use the script with Windows 10/11 unix or Cygwin? What should I put it in?

**Reply by fmw42** [reply]
You open a Unix window or a Cygwin window.  I am sorry, but I am not a Windows user. So I cannot advise on how to use Windows.
